### PR TITLE
Add merge_env/1 and merge_env/2

### PR DESCRIPTION
**Pull request by josevalim**
This makes it simpler and faster to set the configuration of multiple
applications at once, which improves other configuration engines
such as cuttlefish and Elixir's `config.exs`.

While it is possible to achieve the same behaviour by calling
`application:set_env/4` in a loop, such approach requires a lot
of back and forth between the client and the application controller
and having a specialized function speeds up the process.

### CODE REVIEW

File: lib/kernel/src/application_controller.erl [hunk 1]
```diff
@@ -1684,10 +1700,7 @@ get_env_key([H | T], Key, Res) ->
 get_env_key([], _Key, Res) -> Res.
 
 add_env(Name, Env) ->
-    foreach(fun({Key, Value}) ->
-			  ets:insert(ac_tab, {{env, Name, Key}, Value})
-		  end,
-		  Env).
+    ets:insert(ac_tab, [{{env, Name, Key}, Value} || {Key, Value} <- Env]).
```
Comment (fenollp) [inline lib/kernel/src/application_controller.erl]: Not sure if this is relevant here but note that this no longer crashes on lists that contain something else than 2-tuples.

Comment (josevalim) [reply lib/kernel/src/application_controller.erl]: Yup, I checked the call sites and they all validate the data before, so we are good!

**Comment by sirihansen** [comment]
Hi, I think this functionality is good, but I find the new function name a bit confusing. I think it could be better to call it `set_env`, which will also indicate that it does more or less the same as set_env/3,4. What do you think?

To include the PR, I would also need more documentation:
- Explain that this is equivalent to calling set_env/3,4 multiple times, except it has better performance
- Explain what happens if an App/Key occurs several times

And please add tests.

**Comment by sirihansen** [comment]
That's a good point about `get_all_env`, but I also think that `set_all_env` would be a bit confusing - since what would it mean - would it unset envs that were already set but not mentioned? In that case maybe `set_some_env` would be more correct, but I don't think that would add much to the understanding. Anyway - if you're good with `set_env/1,2`, then let's do that :) I'll just wait for your updates then!

**Comment by josevalim** [comment]
I am not sure I follow. :) what other sanity checks are there? We do check
for the list format but that’s because we are passing prop lists around
which is not the case with set_env.
-- 


*José Valimwww.plataformatec.com.br
<http://www.plataformatec.com.br/>Founder and Director of R&D*

**Comment by sirihansen** [comment]
From your function you hit `application_controller:check_conf_data/1`, which is (sadly) not the case from `set_env/3,4`.

**Comment by sirihansen** [comment]
My only point being that your function is not (as I earlier request to be documented) equivalent to calling set_env/3,4 multiple times.

**Comment by sirihansen** [comment]
It is similar, but more efficient + adds sanity check :-D

**Comment by josevalim** [comment]
@sirihansen so I have found some inconsistencies on how config files handle last entries.

Take these three config files:

```erlang
% kernel.config
[{kernel, []}].

% dupkeys.config
[{kernel, [{key1, foo}, {key1, bar}]}].

% dupapps.config
[{kernel, [{key1, foo}]}, {kernel, [{key1, bar}]}].
```

Here are the results for different invocations:

```
$ erl -noinput -config dupapps.config -eval "erlang:display(application:get_env(kernel, key1))." -s erlang halt
{ok,foo}
```

```
$ erl -noinput -config dupkeys.config -eval "erlang:display(application:get_env(kernel, key1))." -s erlang halt
{ok,foo}
```

```
$ erl -noinput -config kernel.config -config dupkeys.config -eval "erlang:display(application:get_env(kernel, key1))." -s erlang halt
{ok,foo}
```

```
$ erl -noinput -config kernel.config -config dupapps.config -eval "erlang:display(application:get_env(kernel, key1))." -s erlang halt
{ok,bar}
```

If we have duplicate entries, we usually stick with the first one unless a previous value was set to that application and we have duplicate application configs. Looking at the code, it seems it was not designed to take duplicate entries into account, so the behaviour is rather unspecified. I couldn't see it documented either.

To further complicate matters, application:load/1 always keeps the last entry for duplicate keys in the application environment.

I can think of a couple options here:

  1. Clearly document on all of those operations that what happens when there are duplicate keys is unspecified. I am not a big fan of this approach but it would be backwards compatible.

  2. Unify all of those to keep the last or the first value - first value seems to be the least breaking change to do.

Thoughts?

**Comment by okeuday** [comment]
@josevalim A way of describing it should be the first value of the last application configuration (having the same application mentioned twice in the config file is similar to using two separate config files for the same application in the same order).  So, it seems best to have application env loading fail when duplicate keys are present, to avoid misconfiguration IMO.  However, a change like that would probably want to be in a separate pull request, to handle both .app files and config files (reltool could check .app files, though reltool doesn't seem like it would have a config file validation option currently).

**Comment by josevalim** [comment]
> @josevalim A way of describing it should be the first value of the last application configuration 

I believe this breaks your description:

> $ erl -noinput -config dupapps.config -eval "erlang:display(application:get_env(kernel, key1))." -s erlang halt
{ok,foo}

We have the application twice and the first value of the first application wins. So the application appearing twice in the same file is not the same as having the application in two separate config files.

**Comment by okeuday** [comment]
@josevalim Yes, it is a bit weirder, it only considers the first entry for an application in a config file and only uses the first key (if the key is duplicated) in the application's config.  Avoiding duplicated configuration by blocking execution (failing fast) would be very helpful.  I have seen complex configurations in the past and have used ad-hoc file:consult/1 use to catch problems manually, but they are easy to accumulate if nothing prevents configuration problems.

**Comment by sirihansen** [comment]
Oh joy! Why did we open that door? ;-)

I am not sure what to do, so just throwing out some thoughts right now.

Would it be a good thing to do a sort/merge thing in your new function before sending the list into the existing code? And document what happens for duplicates. The sort/merge thing could possibly be done in such a way that the function actually becomes equivalent to calling set_env/3,4 in a loop.

Or do you find it better to say that it is equivalent to using configuration files - and then do one of your options 1 or 2 from [this](https://github.com/erlang/otp/pull/2075#issuecomment-454810165) comment?

Regarding the options 1 or 2, I do find it risky to do incompatible changes in this area. But then again, consistency and clear documentation, and a big warning in the release note, still sounds tempting. Does @IngelaAndin , @RaimoNiskanen or @KennethL have any comments on this?

**Comment by okeuday** [comment]
@sirihansen I was hoping extra validation of the config file and the application env configuration could be added to enforce no duplicates.  That would help configuration avoid human error that leads to incorrect runtime.  Then the `merge_env` source code shouldn't have problems (it would be option 2, always taking the first but without duplicate env keys being a possibility).

If validation was added, it would need to cause BEAM to exit if a config had a duplicate application.  The application env duplicate key would cause a failure of the application load (that would cause the BEAM to exit if it occurs via the boot script).  Is there a reason this validation is unable to be added?

**Comment by sirihansen** [comment]
Ok, there are a lot of strange things. Comments related to the merge code indicate that the last occurrence should always take, and in many cases that's what happens. But the code does not expect duplicates of an application within the same file, and neither of the same key within one key-value list. (I also think there might be some missing lists:reverse of `Rest` lists that are build by picking head from one list and inserting as head into another list, in recursion.)

There is also a dependency towards the name of the config file, for instance, a config file containing

```
[{kernel,[{key1,val0},
              {key1,val1}]},
 {kernel,[{key1,val2}]}].
```
gives the value `val1` if it is named `sys.config`, and `val0` if it is named something else. It does not hit `val2` in any case. While this:

```
[{kernel,[{key1,val1}]},
 {kernel,[{key1,val2}]}].
```
gives `val2` if the file is named `sys.config` and `val1` if it is named something else.

So, I think we need to clean this up thoroughly, even if is not backwards compatible. I hope, and believe, that situations where the behavior will change by this correction are few (and mainly caused by unintended configuration content). What do you think?

**Comment by josevalim** [comment]
@sirihansen just to make sure I understood it: the behavioir changes if the file is named sys.config or not? If so, it is preferable to standardize on sys.config, correct? And what happens to config files that are added via sys.config, do they follow the sys.config "convention"?

**Comment by sirihansen** [comment]
Yes, it differs depending on if the name is sys.config or not. The reason is that if it is sys.config, then it is allowed to point out other config files. And this possibility makes the merge take a bit of a different way through the code (multiple calls to merge_env/2 instead of only one, lists are reversed and not reversed back, ...).

I'm not sure if there is a difference between files included via sys.config and files added with and extra `-config` on the command line.

**Comment by josevalim** [comment]
Ok. Here are further tests (using the files from https://github.com/erlang/otp/pull/2075#issuecomment-454810165): 

1. sys.config with duplicate keys

```iex
$ cat sys.config
[{kernel, [{key1, foo}, {key1, bar}]}].
$ erl -noinput -config sys.config -eval "erlang:display(application:get_env(kernel, key1))." -s erlang halt
{ok,foo}
```

2. `sys.config` with duplicate apps:

```iex
$ cat sys.config
[{kernel, [{key1, foo}]}, {kernel, [{key1, bar}]}].
$ erl -noinput -config sys.config -eval "erlang:display(application:get_env(kernel, key1))." -s erlang halt
{ok,bar}
```

3. sys.config with only reference to dup keys

```iex
$ cat sys.config
["dupkeys.config"].
$ erl -noinput -config sys.config -eval "erlang:display(application:get_env(kernel, key1))." -s erlang halt
{ok,foo}
```

4. sys.config with only reference to dup apps

```iex
$ cat sys.config
["dupapps.config"].
$ erl -noinput -config sys.config -eval "erlang:display(application:get_env(kernel, key1))." -s erlang halt
{ok,foo}
```

5. sys.config with empty kernel and reference to dup keys

```iex
$ cat sys.config
[{kernel, []}, "dupkeys.config"].
$ erl -noinput -config sys.config -eval "erlang:display(application:get_env(kernel, key1))." -s erlang halt
{ok,foo}
```

6. sys.config with empty kernel and reference to dup apps

```iex
$ cat sys.config
[{kernel, []}, "dupapps.config"].
$ erl -noinput -config sys.config -eval "erlang:display(application:get_env(kernel, key1))." -s erlang halt
{ok,bar}
```

7. sys.config with kernel and key1 set and reference to dup keys

```iex
$ cat sys.config
[{kernel, [{key1, original}]}, "dupkeys.config"].
$ erl -noinput -config sys.config -eval "erlang:display(application:get_env(kernel, key1))." -s erlang halt
{ok,bar}
```

8. sys.config with kernel and key1 set only reference to dup apps

```iex
$ cat sys.config
[{kernel, [{key1, original}]}, "dupapps.config"].
$ erl -noinput -config sys.config -eval "erlang:display(application:get_env(kernel, key1))." -s erlang halt
{ok,bar}
```

To sum up the sys.config behaviour:

1. When we have duplicate keys, we always return the first key unless the key has already been set, which we return the latest one.

2. When we have duplicate apps, the last one always wins except if there is a previous configuration file and the application has not been set yet.

**Comment by josevalim** [comment]
If we want to do the minimum amount of changes, here is what should be proposed:

### Rules for keys

The first key always wins. EXCEPT under this scenario, which should be considered a bug:

  * a previous value for said key was set in a previous config file.

### Rules for apps

The second application always wins AS LONG AS:

  * it happens inside sys.log; OR
  * it happens in a separate config file and a previous config file set any value for said application

All other scenarios should be fixed.

### Summary

This behaviour only happens if we have duplicate apps and duplicate keys. If there are no duplicate keys or duplicate apps, everything works as expected.

My suggestion right now is to possibly emit a warning and keep the same behaviour. In the future, we can either error out or choose a unified behaviour.

**Comment by sirihansen** [comment]
Thanks for the extended tests! Just to change the perspective for a while - what would be the desired behavior if we didn't have to consider backwards compatibility? I'm thinking:

1. File name should not matter (except for the fact that only sys.config may include other config files).
2. Duplicate keys in the same file, no matter if it is within the same key-value list or in different app-items, should be considered a mistake, and thus not be allowed.
3. Duplicate keys in different files should be ok, and the value from the last file should apply.
4. (Duplicate app-items in the same file may be ok, as long as the same key does not exist more than once.)

What do you think about this?

It does of course break all your tests,  @josevalim , since dupapps.config and dupkeys.config both break statement 2.. But if this is acceptable (maybe introduced as doc+runtime warning in 21.3, and error in 22.0, or is this too quick?), then statement 3. (and 4.) actually describe the current behavior.

**Comment by josevalim** [comment]
@sirihansen I agree with 1, 2, and 3 but I would forbid 4 too. Is there any reason to avoid duplicate apps in the same file?

**Comment by okeuday** [comment]
@sirihansen I agree with @josevalim that 1, 2, and 3 are good, though 4 should cause a failure because it allows human error during configuration that would otherwise not be caught and corrected.  It would be best if the configuration duplicates cause an error after the configuration load (not a warning message).

**Comment by sirihansen** [comment]
@josevalim @okeuday Yes, I agree - 4 should also be forbidden. I think I just added it because I had it in one of my tests and it works :-/  We'll do:

1. File name should not matter (except for the fact that only sys.config may include other config files).
2. Duplicate keys in the same file, no matter if it is within the same key-value list or in different app-items, should be considered a mistake, and thus not be allowed.
3. Duplicate keys in different files should be ok, and the value from the last file should apply.
4. Duplicate app-items in the same file should not be allowed

Any thoughts on introducing doc + warning in 21.3, and then change the warning to error in 22.0?

@KennethL , would this be acceptable?

**Comment by josevalim** [comment]
@sirihansen I can start working on this meanwhile but I do have a couple questions:

1. Is there a way to capture standard_input or standard_error and assert on its contents using `ct`?

2. Why the behaviour of allowing other .config files is restricted to sys.config? Wouldn't it be better to generalize it?

**Comment by sirihansen** [comment]
@josevalim Great!

1. Yes, have a look at [ct:capture_start/0](http://erlang.org/doc/man/ct.html#capture_start-0) and friends.
2. I'm not really sure, but maybe it is to avoid the chance of circular inclusion, and possibly to put pressure on the convention of having one (top) file named sys.config (as the release handling expects). I think there has been some discussions about it earlier - probably on erlang-questions.

**Comment by KennethL** [comment]
I think it is a good suggestion to go for 1,2,3,4 as above. I don't see any reason to be compatible with the way it works now since that is just unexpected and inconsistent. If you have decent configurations without dubplicates the new behaviour will not change anything. If duplicates exist today the only reasonable use of that that I can think of is to override a specific value by setting that in a second config file or in runtime. 
So the disturbance by being incompatible when correcting these bugs should be minial , at least that's what I think.

**Comment by sirihansen** [comment]
@josevalim The two-step approach (warning in 21.3, error in 22.0) is acceptable, but requires the change to be in place at the latest February 25th, to be included in 21.3. Is that possible for you?

**Comment by josevalim** [comment]
Hi @sirihansen, I have sent a PR that fails on duplicate apps/keys here: #2157.

Unfortunately it took me longer than I expected because I struggled with the tests. AFAIK, there are currently no tests for the `check_conf_data` functionality and my attempt at setting up such test suite was flaky and verbose because it required booting the VM multiple times and asserting on the VM output.

The good news is that, if we add `application:set_env/1` and `application:set_env/2`, as outlined in this PR, we will have a mechanism to directly test `check_config_data`. So I can see two paths forward:

1. Merge #2157 without tests and we will get better coverage when this is merged (including for code that exists today and is not covered)

2. Do not merge #2157 and instead I incorporate those changes in this PR, so we get the new code already with proper coverage

I would prefer 2 because I am personally not confident on including changes to the config code without automated tests (of course, the changes have been manually verified).

**Comment by sirihansen** [comment]
@josevalim I also think that 2 is the best option. And please rebase it on maint so we get the warning in OTP-21.3

**Comment by josevalim** [comment]
Closing in favor of #2164
